File: pset2/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("missing command-line argument\n");
        return 1;
    } 
    
    string k = argv[1];
    int klength = strlen(k);
    
    for (int a = 0; a < klength; a++) 
    {
        if (!isalpha(k[a]))
        {
            printf("only enter alphabetical characters\n");
            return 1;
        }
    }
    
    printf("plaintext: ");
    string p = get_string();
    int plength = strlen(p);
    
    printf("ciphertext: ");
    int j = 0;
    for (int i = 0; i < plength; i++)
    {
        if (j >= klength) {
            j = 0;
        }
        if (isalpha(p[i]))
       {
            if (isupper(p[i]))
            {
                if (isupper(k[j]))
                {
                    printf("%c", ((p[i] - 65 + (k[j] - 65)) % 26 + 65));
                } else 
                {
                    printf("%c", ((p[i] - 65 + (k[j] - 97)) % 26 + 65));
                }
            } else if (islower(p[i]))
            {
                if (isupper(k[j]))
                {
                    printf("%c", ((p[i] - 97 + k[j] - 65) % 26 + 97));
                } else
                {
                    printf("%c", ((p[i] - 97 + k[j] - 97) % 26 + 97));
                }
            } 
            j += 1;
        } else 
        { 
            printf("%c", p[i]);
        }
    }
    printf("\n");
}

File: pset6/vigenere.py
import sys
import cs50

def main():
    if (len(sys.argv) != 2):
        print("missing command-line argument")
        exit(1)
        
    # store cipher key
    k = sys.argv[1]
    
    # return 1 if non-alphabetical characters in k
    if (not k.isalpha()):
        print("only enter alphabetical characters")
        exit(1)    
    
    # convert cipher key to rotation number
    r = list(k)
    
    for i in range(len(r)):
        # convert uppercase
        if (r[i].isupper()):
            r[i] = ord(k[i]) - 65
        else:
            r[i] = ord(k[i]) - 97
            
    
    # ask user for plaintext to cipher
    print("plaintext: ", end="")
    p = cs50.get_string()
    print("ciphertext: ", end="")
    # initialise position in rotation number
    rp = 0
    
    # iterate through characters in plaintext
    for char in range(len(p)):
        # if position in cipher key is longer than key, reset to 0 
        if (rp >= len(k)):
            rp = 0
        # check if character is alphabetical
        if (p[char].isalpha()):
            # check if character is uppercase
            if (p[char].isupper()):
                print(chr(((ord(p[char]) - 65 + r[rp]) % 26 + 65)), end="")
            else:
                print(chr(((ord(p[char]) - 97 + r[rp]) % 26 + 97)), end="")
            rp += 1
        else:
            print(p[char], end="")

    print("")

if __name__ == "__main__":
    main()